keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1563 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This enhancement proposal aims at improving the quality of service of
+applications in Kubernetes by introducing a new type of resource control
+mechanism. Certain types of resources are inherently shared by application (e.g.
+cache, memory bandwidth and disk I/O) and while there are technologies for
+controlling these, there is currently no meaningful way in Kubernetes to
+support those tehcnologies. This proposal suggests to address the issue above
+in a generalized way by extending the Kubernetes resource model with a new type
+of resources, i.e. QoS-class resources.
+
+[Intel RDT][intel-rdt] implements a class-based mechanism for controlling the
+cache and memory bandwidth QoS of applications. All processes in the same
+hardware class share a portion of cache lines and memory bandwidth. RDT
+provides a way for mitigating noisy neighbors and fulfilling SLAs. In Linux
+control happens via resctrl -- a pseudo-filesystem provided by the kernel which
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already. Other hardware vendors have
+comparable technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provies a possible framework for the future improvements, for instance enabling
+class-based network or memory type prioritization of applications.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources.
+CRI-O and containerd runtimes have support for RDT and blockio classes and they
+provide an bridge-gap user interface through special pod annotations. We would
+like to eventually get these types of resources first class citizen and
+properly supported in Kubernetes, providing visibility, a well-defined user
+interface, and permission controls.
+
+It seems necessary to support both container-level and pod-level QoS-class
+resources as independent concepts. Intel RDT (above) is per-container by
+design because of the hardware implementation (the control/class hiearchy is
+flat). Also, the current support for blockio is container-level only (it is not
+possible to configure pod sandbox-level cgroup parameters). However, having
+pod-level QoS-class resources makes it possible to implement support for
```

Comment:
Is it worth seperating this part from the feature?

Reply:
Sorry, I didn't quite understand your comment. Separating which part from which feature?

Reply:
> Sorry, I didn't quite understand your comment. Separating which part from which feature?